disperser/meterer/meterer.go

```diff
@@ -0,0 +1,273 @@
+	// }
+
+	// Validate against the payment method
+	if header.CumulativePayment == 0 {
+		reservation, err := m.ChainState.MockedGetActiveReservationByAccount(ctx, blockNumber, header.AccountID)
+		if err != nil {
+			return fmt.Errorf("failed to get active reservation by account: %w", err)
+		}
+		if err := m.ServeReservationRequest(ctx, header, reservation); err != nil {
+			return fmt.Errorf("invalid reservation: %w", err)
+		}
+	} else {
+		onDemandPayment, err := m.ChainState.MockedGetOnDemandPaymentByAccount(ctx, blockNumber, header.AccountID)
+		if err != nil {
+			return fmt.Errorf("failed to get on-demand payment by account: %w", err)
+		}
+		if err := m.ServeOnDemandRequest(ctx, header, onDemandPayment); err != nil {
+			return fmt.Errorf("invalid on-demand request: %w", err)
+		}
+	}
+
+	// // If all checks pass, add the blob header to the meterer's state
+	// if err := m.addToState(ctx, header); err != nil {
+	// 	return fmt.Errorf("failed to add to meterer state: %w", err)
+	// }
+
+	return nil
+}
+
+// TODO: mocked EIP712 domain, change to the real thing when available
+// ValidateSignature checks if the signature is valid against all other fields in the header
+// Assuming the signature is an eip712 signature
+func (m *Meterer) ValidateSignature(ctx context.Context, header BlobHeader) error {
+	// Create the EIP712Signer
+	//TODO: update the chainID and verifyingContract
+	signer := NewEIP712Signer(big.NewInt(17000), common.HexToAddress("0x1234000000000000000000000000000000000000"))
+
+	recoveredAddress, err := signer.RecoverSender(&header)
+	if err != nil {
+		return fmt.Errorf("failed to recover sender: %w", err)
+	}
+
+	accountAddress := common.HexToAddress(header.AccountID)
+
+	if recoveredAddress != accountAddress {
+		return fmt.Errorf("invalid signature: recovered address %s does not match account ID %s", recoveredAddress.Hex(), accountAddress.Hex())
+	}
+
+	return nil
+}
+
+// ServeReservationRequest handles the rate limiting logic for incoming requests
+func (m *Meterer) ServeReservationRequest(ctx context.Context, blobHeader BlobHeader, reservation *ActiveReservation) error {
+	if !m.ValidateBinIndex(blobHeader, reservation) {
+		return fmt.Errorf("invalid bin index for reservation")
+	}
+
+	// Update bin usage atomically and check against reservation's data rate as the bin limit
+	if err := m.IncrementBinUsage(ctx, blobHeader, reservation.DataRate); err != nil {
+		return fmt.Errorf("bin overflows: %w", err)
+	}
+
+	return nil
+}
+
+// ValidateBinIndex checks if the provided bin index is valid
+func (m *Meterer) ValidateBinIndex(blobHeader BlobHeader, reservation *ActiveReservation) bool {
+	currentBinIndex := GetCurrentBinIndex()
+	// Valid bin indexes are either the current bin or the previous bin
+	if (blobHeader.BinIndex != currentBinIndex && blobHeader.BinIndex != (currentBinIndex-1)) || (reservation.StartEpoch > blobHeader.BinIndex || blobHeader.BinIndex > reservation.EndEpoch) {
```

Comment:
Does it make sense for Epochs to use the same units as Bins?

Reply:
I'm open to change this for easier usage. What about using start and end timestamps instead of epochs?